README.rst

```diff
@@ -8,5 +8,195 @@
+  this package is integrated with the `transaction
+  <https://transaction.readthedocs.io>`_ package.
+
+  Resources are externalized during a late phase of the transaction
+  commit process; the details about the delivery are recorded and
+  persisted, and only after the transaction is successfully committed
+  does the HTTP request get made.
+
+- Concurrency
+
+  Webhook delivery and record keeping should be lightweight, and
+  all actual network IO should proceed in a non-blocking fashion. This
+  means that this package will spawn threads (or greenlets, using
+  `gevent <http://www.gevent.org>`_.
+
+- Error Handling/Failure Retry
+
+  A limited amount of retry logic is provided by this package, but
+  that does not extend to process boundaries. If the process hosting
+  this package is killed while a delivery is pending, no automatic
+  provision is made to resume delivery attempts in any other process.
+
+  The API is present to allow that to be implemented, though.
+
+- Auditing/Delivery History
+
+  For each subscription, delivery attempts, status, and responses are
+  stored in a ring-buffer like structure. This can be inspected to see
+  if deliveries succeeded, failed, or never completed.
+
+- Access Control on Deliveries
+
+  Each subscription is associated with an ``IPrincipal`` that owns it.
+  A request is only delivered to a subscription if the ``IPrincipal``
+  that owns the subscription can access the entity, as determined by
+  `zope.security <https://zopesecurity.readthedocs.io>`_.
+
+- Access Control on Subscriptions
+
+  While not enforced by this package, the above owner relationship
+  will be used to provide role managers that grant read and read/write
+  access to remove subscriptions only to the owner of the
+  subscription.
+
+  TODO: Make sure client packages can extend that to provide for admin
+  access. So long as we don't DENY it should be fine.
+
+- Hierarchy of Subscriptions
+
+  Subscriptions are made within a particular Zope site (the closest
+  enclosing site to a resource when a resource is subscribed to, or
+  the currently active site otherwise). These sites may have parents.
+
+  TODO: Work out the details of that.
+
+  When an event is received that might result in webhook delivery,
+  active subscriptions are checked for in the currently active site,
+  as well as in the sites up the hierarchy of the resource itself. All
+  applicable subscribers will get a delivery.
+
+  For example, if the president of the company (an administrator)
+  subscribes to "new user created" events at the global (root, base or
+  "/") level, and a department head subscribes to "new user created"
+  for their department ("/NOAA"), while a local office manager
+  subscribes to events for their office ("/NOAA/NWS/OKC"), then
+  creating a new user in the OKC office may send three deliveries, one
+  to the manager, one to the secretary, and one to the president.
```

Comment:
💯 Nice. Super powerful and I think that will be important for enterprise installations